plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseCaseInsensitiveMappingTest.java

```diff
@@ -311,6 +312,65 @@ public void testSchemaAndTableNameRuleMapping()
+            assertTableColumnNames("remote_schema.remote_table", "remote_column");
+            assertQuery("SHOW COLUMNS FROM remote_schema.remote_table", "SELECT 'remote_column', 'varchar(5)', '', ''");
+            assertUpdate("INSERT INTO remote_schema.remote_table VALUES 'grant'", 1);
+            assertUpdate("INSERT INTO remote_schema.remote_table (remote_column) VALUES 'santa'", 1);
+            assertQueryFails("INSERT INTO remote_schema.remote_table (remoteColumn) VALUES 'athing'", "Insert column name does not exist in target table: remotecolumn");
+            assertQuery("SELECT * FROM remote_schema.remote_table", "VALUES 'grant', 'santa'");
+            assertQuery("SELECT remote_column FROM remote_schema.remote_table", "VALUES 'grant', 'santa'");
+
+            assertUpdate("CREATE TABLE remote_schema.ctas as select remote_column from remote_schema.remote_table", 2);
+            assertQuery("SELECT remote_column FROM remote_schema.ctas", "VALUES 'grant', 'santa'");
+            assertUpdate("DELETE FROM remote_schema.ctas", 2);
+
+            assertUpdate("CREATE TABLE remote_schema.ctas_star as select * from remote_schema.remote_table", 2);
+            assertQuery("SELECT remote_column FROM remote_schema.ctas_star", "VALUES 'grant', 'santa'");
+        }
+    }
+
+    @Test
+    public void testSchemaAndTableMappingsWithColumnMappings()
+            throws Exception
+    {
+        updateRuleBasedIdentifierMappingFile(
+                getMappingFile(),
+                ImmutableList.of(new SchemaMappingRule("RemoteSchema", "remote_schema")),
+                ImmutableList.of(new TableMappingRule("RemoteSchema", "RemoteTable", "remote_table")),
+                ImmutableList.of(
+                        new ColumnMappingRule("RemoteSchema", "RemoteTable", "RemoteColumn", "remote_column"),
+                        new ColumnMappingRule("RemoteSchema", "RemoteTable", "Casesensitive", "caseinsensitive")));
+
+        try (AutoCloseable ignore1 = withSchema("RemoteSchema");
+                AutoCloseable ignore2 = withTable("RemoteSchema", "RemoteTable", "(Casesensitive varchar(5), RemoteColumn int)")) {
+            assertTableColumnNames("remote_schema.remote_table", "caseinsensitive", "remote_column");
+            assertUpdate("INSERT INTO remote_schema.remote_table VALUES ('grant', 20)", 1);
+            assertUpdate("INSERT INTO remote_schema.remote_table (caseinsensitive, remote_column) VALUES ('santa', 100)", 1);
+            assertUpdate("INSERT INTO remote_schema.remote_table (caseinsensitive) VALUES ('test')", 1);
+
+            assertQuery("SELECT * FROM remote_schema.remote_table", "VALUES ('grant', 20), ('santa', 100), ('test', null)");
+            assertQuery("SELECT caseinsensitive, remote_column FROM remote_schema.remote_table", "VALUES ('grant', 20), ('santa', 100), ('test', null)");
+            assertUpdate("CREATE TABLE remote_schema.ctas as select * from remote_schema.remote_table", 3);
```

Comment:
It has been hard to find all the edge cases (insert, select, delete, CTAS, using * or named columns, etc) and I'm still testing.

The problem is it is very easy to use the "wrong" column name because both remote/mapped names are Strings. If we wrapped remote/mapped names in separate types, it might help catch a lot of these errors at compile time.